src/handlers/graphical.rs

```diff
@@ -377,66 +377,63 @@ impl GraphicalReportHandler {
-                                        self.context_lines,
-                                    )
-                                    .is_ok()
-                                {
-                                    contexts.pop();
-                                    contexts.push((
-                                        // We'll throw this away later
-                                        new_span, left_conts,
-                                    ));
-                                } else {
-                                    contexts.push((right, right_conts));
-                                }
-                            } else {
-                                contexts.push((right, right_conts));
-                            }
-                        }
-                    }
-                    for (ctx, _) in contexts {
-                        self.render_context(f, source, &ctx, &labels[..])?;
-                    }
+        let Some(source) = opt_source else {
+            return Ok(());
+        };
+        let Some(labels) = diagnostic.labels() else {
+            return Ok(());
+        };
+
+        let mut labels = labels.collect::<Vec<_>>();
+        labels.sort_unstable_by_key(|l| l.inner().offset());
+
+        let mut contexts = Vec::with_capacity(labels.len());
+        for right in labels.iter().cloned() {
+            let right_conts = source
+                .read_span(right.inner(), self.context_lines, self.context_lines)
+                .map_err(|_| fmt::Error)?;
+
+            if contexts.is_empty() {
+                contexts.push((right, (right_conts.line(), right_conts.line_count())));
+                continue;
+            }
+
+            let (left, (left_line, left_line_count)) = contexts.last().unwrap().clone();
+            if left_line + left_line_count >= right_conts.line() {
+                // The snippets will overlap, so we create one Big Chunky Boi
+                let left_end = left.offset() + left.len();
+                let right_end = right.offset() + right.len();
+                let new_span = LabeledSpan::new(
+                    left.label().map(String::from),
+                    left.offset(),
+                    if right_end >= left_end {
+                        // Right end goes past left end
+                        right_end - left.offset()
+                    } else {
+                        // right is contained inside left
+                        left.len()
+                    },
+                );
+                // Check that the two contexts can be combined
+                if let Ok(new_conts) =
+                    source.read_span(new_span.inner(), self.context_lines, self.context_lines)
+                {
+                    contexts.pop();
+                    contexts.push((
+                        // We'll throw this away later
+                        new_span,
+                        (new_conts.line(), new_conts.line_count()),
```

Comment:
We return the combined context lines here instead of only the context lines of only the first label to fix the bug mentioned in the PR description.
(`new_conts` contains the combined context lines and `left_conts` contains only context lines from the first label)